traffic_ops/traffic_ops_golang/deliveryservice/consistenthash/consistenthash.go

```diff
@@ -0,0 +1,131 @@
+package consistenthash
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+// struct for the response object from Traffic Router
+type TRConsistentHashResult struct {
+	ResultingPathToConsistentHash string `json:"resultingPathToConsistentHash"`
+	ConsistentHashRegex           string `json:"consistentHashRegex"`
+	RequestPath                   string `json:"requestPath"`
+}
+
+// struct for the incoming request object
+type TRConsistentHashRequest struct {
+	ConsistentHashRegex string `json:"regex"`
+	RequestPath         string `json:"requestPath"`
+	CdnID               int64  `json:"cdnId"`
+}
+
+// endpoint to test Traffic Router's Pattern-Based Consistent Hashing feature
+// /api/1.5/consistenthash
+func Post(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	req := TRConsistentHashRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("malformed JSON: "+err.Error()), nil)
+		return
+	}
+
+	responseFromTR, err := getPatternBasedConsistentHash(inf.Tx.Tx, req.ConsistentHashRegex, req.RequestPath, req.CdnID)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting pattern based consistent hash from Traffic Router: "+err.Error()))
+		return
+	}
+
+	consistentHashResult := TRConsistentHashResult{}
+	json.Unmarshal(responseFromTR, &consistentHashResult)
+
+	api.WriteResp(w, r, consistentHashResult)
+}
+
+const RouterRequestTimeout = time.Second * 10
+
+// queries database for active Traffic Router on the CDN specified by cdnId
+// passes regex and requestPath to the Traffic Router via API request,
+// and returns the response
+func getPatternBasedConsistentHash(tx *sql.Tx, regex string, requestPath string, cdnId int64) ([]byte, error) {
+	q := `
```

Comment:
I'm not clear based on existing code what the procedure is for things like this, but could this be made into a `const`? The compiler might be smart enough to handle it, but allocating this space on a per-call basis may be a waste of cycles and threaded memory.

Reply:
this follows the existing procedure for database queries. also, this isn't an endpoint that will be getting much traffic.

Reply:
Yeah we do this same thing all over the place in traffic_ops_golang. Also, premature optimization is the root of all evil.

Reply:
I'm fine leaving it the way it is, but on that note: why are things this way? Is there an advantage to e.g. creating a `func` that does nothing but return a static, constant string?

Reply:
Readability. If it really is a constant string, there's no reason to do a `func` instead of a `const`. Unless maybe ever endpoint does `func selectQuery`, and then the consistency is easier to read. But some things can't be `const` in Go (e.g. `map`). Then, a `func` makes it impossible to modify, and is safer than a global `var`.

Here, it's arguably easier to read:
```
qry := ` 
// seven lines 
`
tx.Exec(qry, cdnID)
```
Than: ```
tx.Exec(`
// seven lines
, cdnID)
```

>allocating this space on a per-call basis may be a waste of cycles and threaded memory.

The compiler is smarter than you and me. Are you certain it's faster to reference a distant heap object every usage, than to copy the memory onto the stack and reference local memory? I'm not. 

Even the compiler isn't smart enough, things should only be optimized when necessary; here it's not. Readability is more important.

"Programs must be written for people to read, and only incidentally for machines to execute." —Abelson, SICP